File: backend/src/FieldTicket.Infrastructure/Services/CommunicationTemplateService.cs
using FieldTicket.Core.Services;
using FieldTicket.Domain.Entities;
using FieldTicket.Infrastructure.Data;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

namespace FieldTicket.Infrastructure.Services;

/// <summary>
/// 客户沟通模板服务实现
/// </summary>
public class CommunicationTemplateService : ICommunicationTemplateService
{
    private readonly ApplicationDbContext _dbContext;
    private readonly ILogger<CommunicationTemplateService> _logger;

    public CommunicationTemplateService(
        ApplicationDbContext dbContext,
        ILogger<CommunicationTemplateService> logger)
    {
        _dbContext = dbContext;
        _logger = logger;
    }

    public async Task<List<CommunicationTemplateDto>> GetTemplatesAsync(string? scenario = null)
    {
        var query = _dbContext.CommunicationTemplates
            .Where(t => t.IsActive);

        if (!string.IsNullOrEmpty(scenario))
        {
            query = query.Where(t => t.Scenario == scenario);
        }

        var templates = await query
            .OrderBy(t => t.Scenario)
            .ThenBy(t => t.Name)
            .ToListAsync();

        return templates.Select(MapToDto).ToList();
    }

    public async Task<CommunicationTemplateDto?> GetTemplateAsync(Guid templateId)
    {
        var template = await _dbContext.CommunicationTemplates
            .FirstOrDefaultAsync(t => t.TemplateId == templateId);

        return template != null ? MapToDto(template) : null;
    }

    public async Task<CommunicationTemplateDto?> GetTemplateByCodeAsync(string code)
    {
        var template = await _dbContext.CommunicationTemplates
            .FirstOrDefaultAsync(t => t.Code == code);

        return template != null ? MapToDto(template) : null;
    }

    public async Task<Guid> CreateTemplateAsync(CreateCommunicationTemplateRequest request, Guid userId)
    {
        var template = new CommunicationTemplate
        {
            TemplateId = Guid.NewGuid(),
            Code = request.Code,
            Name = request.Name,
            Scenario = request.Scenario,
            ContentTemplate = request.ContentTemplate,
            IsActive = request.IsActive,
            CreatedBy = userId,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        _dbContext.CommunicationTemplates.Add(template);
        await _dbContext.SaveChangesAsync();

        _logger.LogInformation("创建沟通模板：{Code} - {Name}", request.Code, request.Name);

        return template.TemplateId;
    }

    public async Task UpdateTemplateAsync(Guid templateId, UpdateCommunicationTemplateRequest request)
    {
        var template = await _dbContext.CommunicationTemplates
            .FirstOrDefaultAsync(t => t.TemplateId == templateId);

        if (template == null)
        {
            throw new KeyNotFoundException($"模板 {templateId} 不存在");
        }

        if (request.Name != null)
        {
            template.Name = request.Name;
        }
        if (request.Scenario != null)
        {
            template.Scenario = request.Scenario;
        }
        if (request.ContentTemplate != null)
        {
            template.ContentTemplate = request.ContentTemplate;
        }
        if (request.IsActive.HasValue)
        {
            template.IsActive = request.IsActive.Value;
        }

        template.UpdatedAt = DateTime.UtcNow;

        await _dbContext.SaveChangesAsync();

        _logger.LogInformation("更新沟通模板：{TemplateId}", templateId);
    }

    public async Task DeleteTemplateAsync(Guid templateId)
    {
        var template = await _dbContext.CommunicationTemplates
            .FirstOrDefaultAsync(t => t.TemplateId == templateId);

        if (template == null)
        {
            throw new KeyNotFoundException($"模板 {templateId} 不存在");
        }

        _dbContext.CommunicationTemplates.Remove(template);
        await _dbContext.SaveChangesAsync();

        _logger.LogInformation("删除沟通模板：{TemplateId}", templateId);
    }

    public async Task<string> RenderTemplateAsync(Guid templateId, Dictionary<string, string> variables)
    {
        var template = await _dbContext.CommunicationTemplates
            .FirstOrDefaultAsync(t => t.TemplateId == templateId);

        if (template == null)
        {
            throw new KeyNotFoundException($"模板 {templateId} 不存在");
        }

        var content = template.ContentTemplate;

        // 替换变量 {variable_name}
        foreach (var variable in variables)
        {
            content = content.Replace($"{{{variable.Key}}}", variable.Value);
        }

        return content;
    }

    public async Task<Guid> SaveCommunicationAsync(CreateCommunicationRequest request, Guid userId)
    {
        // 映射沟通类型（小程序使用 call/message/email/other，后端使用 wechat/phone/email/other）
        var communicationType = request.CommunicationType switch
        {
            "call" => "phone",
            "message" => "wechat",
            "email" => "email",
            "other" => "other",
            _ => request.CommunicationType // 如果已经是后端格式，直接使用
        };

        var communication = new CustomerCommunication
        {
            CommunicationId = Guid.NewGuid(),
            TicketId = request.TicketId,
            TemplateId = request.TemplateId,
            Content = request.Content,
            CommunicationType = communicationType,
            CommunicatedBy = userId,
            CommunicatedAt = DateTime.UtcNow,
            CustomerFeedback = request.CustomerFeedback,
            CreatedAt = DateTime.UtcNow
        };

        _dbContext.CustomerCommunications.Add(communication);
        await _dbContext.SaveChangesAsync();

        _logger.LogInformation("保存沟通记录：工单 {TicketId}，类型 {Type}", request.TicketId, request.CommunicationType);

        return communication.CommunicationId;
    }

    public async Task<List<CustomerCommunicationDto>> GetTicketCommunicationsAsync(Guid ticketId)
    {
        var communications = await _dbContext.CustomerCommunications
            .Where(c => c.TicketId == ticketId)
            .OrderByDescending(c => c.CommunicatedAt)
            .ToListAsync();

        var result = new List<CustomerCommunicationDto>();

        foreach (var comm in communications)
        {
            // 映射沟通类型（后端使用 wechat/phone/email/other，小程序使用 call/message/email/other）
            var communicationType = comm.CommunicationType switch
            {
                "phone" => "call",
                "wechat" => "message",
                "email" => "email",
                "other" => "other",
                _ => comm.CommunicationType
            };

            var dto = new CustomerCommunicationDto
            {
                CommunicationId = comm.CommunicationId,
                TicketId = comm.TicketId,
                TemplateId = comm.TemplateId,
                Content = comm.Content,
                CommunicationType = communicationType,
                CommunicatedBy = comm.CommunicatedBy,
                CommunicatedAt = comm.CommunicatedAt,
                CustomerFeedback = comm.CustomerFeedback
            };

            // 获取模板名称
            if (comm.TemplateId.HasValue)
            {
                var template = await _dbContext.CommunicationTemplates
                    .FirstOrDefaultAsync(t => t.TemplateId == comm.TemplateId.Value);
                if (template != null)
                {
                    dto.TemplateName = template.Name;
                }
            }

            // 获取沟通人名称
            var user = await _dbContext.Users
                .FirstOrDefaultAsync(u => u.Id == comm.CommunicatedBy);
            if (user != null)
            {
                dto.CommunicatedByName = user.Name;
            }

            result.Add(dto);
        }

        return result;
    }

    private CommunicationTemplateDto MapToDto(CommunicationTemplate template)
    {
        return new CommunicationTemplateDto
        {
            TemplateId = template.TemplateId,
            Code = template.Code,
            Name = template.Name,
            Scenario = template.Scenario,
            ContentTemplate = template.ContentTemplate,
            IsActive = template.IsActive,
            CreatedAt = template.CreatedAt,
            UpdatedAt = template.UpdatedAt
        };
    }
}

File: web-admin/src/services/communicationTemplateService.ts
/**
 * 客户沟通模板服务
 */
import { authService } from './authService';

const API_BASE_URL = import.meta.env.VITE_API_URL || 'http://localhost:5001';

export interface CommunicationTemplateDto {
  templateId: string;
  code: string;
  name: string;
  scenario: string;
  contentTemplate: string;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface CreateCommunicationTemplateRequest {
  code: string;
  name: string;
  scenario: string;
  contentTemplate: string;
  isActive?: boolean;
}

export interface UpdateCommunicationTemplateRequest {
  name?: string;
  scenario?: string;
  contentTemplate?: string;
  isActive?: boolean;
}

export interface CreateCommunicationRequest {
  ticketId: string;
  templateId?: string;
  content: string;
  communicationType?: string;
  customerFeedback?: string;
}

export interface CustomerCommunicationDto {
  communicationId: string;
  ticketId: string;
  templateId?: string;
  templateName?: string;
  content: string;
  communicationType: string;
  communicatedBy: string;
  communicatedByName?: string;
  communicatedAt: string;
  customerFeedback?: string;
}

class CommunicationTemplateService {
  /**
   * 获取模板列表
   */
  async getTemplates(scenario?: string): Promise<CommunicationTemplateDto[]> {
    const params = scenario ? `?scenario=${encodeURIComponent(scenario)}` : '';
    const response = await fetch(`${API_BASE_URL}/api/communication-templates${params}`, {
      headers: authService.getAuthHeaders(),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ message: '获取模板列表失败' }));
      throw new Error(error.message || '获取模板列表失败');
    }

    return response.json();
  }

  /**
   * 获取模板详情
   */
  async getTemplate(templateId: string): Promise<CommunicationTemplateDto> {
    const response = await fetch(`${API_BASE_URL}/api/communication-templates/${templateId}`, {
      headers: authService.getAuthHeaders(),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ message: '获取模板详情失败' }));
      throw new Error(error.message || '获取模板详情失败');
    }

    return response.json();
  }

  /**
   * 根据编号获取模板
   */
  async getTemplateByCode(code: string): Promise<CommunicationTemplateDto> {
    const response = await fetch(`${API_BASE_URL}/api/communication-templates/code/${encodeURIComponent(code)}`, {
      headers: authService.getAuthHeaders(),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ message: '获取模板失败' }));
      throw new Error(error.message || '获取模板失败');
    }

    return response.json();
  }

  /**
   * 创建模板
   */
  async createTemplate(request: CreateCommunicationTemplateRequest): Promise<string> {
    const response = await fetch(`${API_BASE_URL}/api/communication-templates`, {
      method: 'POST',
      headers: {
        ...authService.getAuthHeaders(),
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ message: '创建模板失败' }));
      throw new Error(error.message || '创建模板失败');
    }

    const result = await response.json();
    return result.templateId;
  }

  /**
   * 更新模板
   */
  async updateTemplate(templateId: string, request: UpdateCommunicationTemplateRequest): Promise<void> {
    const response = await fetch(`${API_BASE_URL}/api/communication-templates/${templateId}`, {
      method: 'PUT',
      headers: {
        ...authService.getAuthHeaders(),
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ message: '更新模板失败' }));
      throw new Error(error.message || '更新模板失败');
    }
  }

  /**
   * 删除模板
   */
  async deleteTemplate(templateId: string): Promise<void> {
    const response = await fetch(`${API_BASE_URL}/api/communication-templates/${templateId}`, {
      method: 'DELETE',
      headers: authService.getAuthHeaders(),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ message: '删除模板失败' }));
      throw new Error(error.message || '删除模板失败');
    }
  }

  /**
   * 渲染模板（替换变量）
   */
  async renderTemplate(templateId: string, variables: Record<string, string>): Promise<string> {
    const response = await fetch(`${API_BASE_URL}/api/communication-templates/${templateId}/render`, {
      method: 'POST',
      headers: {
        ...authService.getAuthHeaders(),
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(variables),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ message: '渲染模板失败' }));
      throw new Error(error.message || '渲染模板失败');
    }

    const result = await response.json();
    return result.content;
  }

  /**
   * 保存沟通记录
   */
  async saveCommunication(request: CreateCommunicationRequest): Promise<string> {
    const response = await fetch(`${API_BASE_URL}/api/communication-templates/communications`, {
      method: 'POST',
      headers: {
        ...authService.getAuthHeaders(),
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ message: '保存沟通记录失败' }));
      throw new Error(error.message || '保存沟通记录失败');
    }

    const result = await response.json();
    return result.communicationId;
  }

  /**
   * 获取工单的沟通记录
   */
  async getTicketCommunications(ticketId: string): Promise<CustomerCommunicationDto[]> {
    const response = await fetch(`${API_BASE_URL}/api/communication-templates/tickets/${ticketId}/communications`, {
      headers: authService.getAuthHeaders(),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ message: '获取沟通记录失败' }));
      throw new Error(error.message || '获取沟通记录失败');
    }

    return response.json();
  }
}

export const communicationTemplateService = new CommunicationTemplateService();


